stdlib/public/core/StringCore2.swift

```diff
@@ -0,0 +1,1061 @@
+    ) {
+      _content = .unowned8(x)
+    }
+    else {
+      _content = .latin1(.copying(source, isASCII: isASCII))
+    }
+  }
+  
+  init(
+    unowned source: UnsafeBufferPointer<UInt16>,
+    isASCII: Bool?,
+    isNULTerminated: Bool
+  ) {
+    if let x = String._Content._Inline<UInt8>(source) {
+      _content = .inline8(x)
+    }
+    else if let x = String._Content._Inline<UInt16>(source) {
+      _content = .inline16(x)
+    }
+    else if let x = String._Content._Unowned<UInt16>(
+      source, isASCII: isASCII,
+      isNULTerminated: isNULTerminated
+    ) {
+      _content = .unowned16(x)
+    }
+    else if isASCII == true || !source.contains { $0 > 0xFF } {
+      _content = .latin1(
+        .copying(_MapCollection(source, through: _TruncExt()), isASCII: true))
+    }
+    else {
+      _content = .utf16(.copying(source))
+    }
+  }
+  
+  public var startIndex: Int { return 0 }
+  public var endIndex: Int { return count }
+  public var count: Int {
+    @inline(__always)
+    get {
+      switch self._content {
+      case .inline8(let x): return x.count
+      case .inline16(let x): return x.count 
+      case .unowned8(let x): return Int(x._count) 
+      case .unowned16(let x): return Int(x._count) 
+      case .latin1(let x):  return x.count 
+      case .utf16(let x): return x.count 
+      case .nsString(let x): return x.length() 
+      }
+    }
+  }
+  
+  public subscript(i: Int) -> UInt16 {
+    @inline(__always)
+    get {
+      switch self._content {
+      case .inline8(var x): return UInt16(x[i])
+      case .inline16(var x):  return x[i]
+      case .unowned8(let x): return UInt16(x.unsafeBuffer[i])
+      case .unowned16(let x): return x.unsafeBuffer[i]
+      case .latin1(let x): return UInt16(x[i])
+      case .utf16(let x): return x[i]
+      case .nsString(let x):
+        return x.characterAtIndex(i)
+      }
+    }
+  }
+
+  public func index(after i: Int) -> Int { return i + 1 }
+  public func index(before i: Int) -> Int { return i - 1 }
+}
+
+extension String._Testable.UTF16View : RangeReplaceableCollection {
+  public var capacity: Int {
+    get {
+      switch self._content {
+      case .inline8(let x): return x.capacity
+      case .inline16(let x): return x.capacity
+      case .unowned8(let x): return Int(x._count)
+      case .unowned16(let x): return Int(x._count)
+      case .latin1(let x): return x.capacity
+      case .utf16(let x): return x.capacity
+      case .nsString(let x): return x.length()
+      }
+    }
+  }
+  
+  public init() {
+    _content = String._Content()
+  }
+
+  internal var _rangeReplaceableStorageID: ObjectIdentifier? {
+    switch self._content {
+    case .latin1(let x): return ObjectIdentifier(x)
+    case .utf16(let x): return ObjectIdentifier(x)
+    default: return nil
+    }
+  }
+
+  internal var _dynamicStorageIsMutable: Bool? {
+    mutating get {
+      return _rangeReplaceableStorageID?._liveObjectIsUniquelyReferenced()
+    }
+  }
```

Comment:
Ah, yes.  Well, indirect cases should be automatically COW'd anyway, eventually.